### PR TITLE
[#159][Part 3][Template] Support datastore replacing shared preference

### LINKED ISSUE #159: Support DataStore in a replacement for SharedPreference
## Why

As recommended by Google, it is more efficient and practical to use DataStore instead of SharedPreference. Perhaps it is time for us to consider having this either as a mandatory or an option to initialize a project.

Read more:
- https://android-developers.googleblog.com/2020/09/prefer-storing-data-with-jetpack.html

- https://developer.android.com/topic/libraries/architecture/datastore

## Who Benefits?

Template users.

👍 : make it a default choice of storage
👎  : make it an optional choice after SharedPreference.

### PR DESCRIPTION
**Pull request by lydiasama**
- Closes #159 

## What happened 👀

- [x] Part 1: Integrate for sample-compose
- [x] Part 2: Integrate for sample-xml
- [x] Part 3: Integrate for both template-xml & template-compose

## Insight 📝

- Add the DataStore Preferences dependencies to the `Template-Xml` and `Template-Compose` projects.
- Add only dependencies, don't add the use cases for template repos as we already had the use case examples in the sample repos.
- Used the same dependencies and logic implementation from #389 

## Proof Of Work 📹

All unit tests passed.

### CODE REVIEW

File: template-xml/app/src/main/java/co/nimblehq/template/xml/di/modules/StorageModule.kt [hunk 1]
```diff
@@ -17,18 +14,8 @@ abstract class StorageModule {
...
-
         @Provides
         fun provideSecuredLocalStorage(@ApplicationContext context: Context) =
             EncryptedSharedPreferences(context)
-
-        @Provides
-        fun provideNormalLocalStorage(
-            @ApplicationContext context: Context,
-            defaultSharedPreferences: SharedPreferences
-        ) = NormalSharedPreferences(context, defaultSharedPreferences)
```
Comment (luongvo) [inline template-xml/app/src/main/java/co/nimblehq/template/xml/di/modules/StorageModule.kt:32]: I think we forgot to remove deprecated NormalSharedPreferences

File: template-compose/app/src/main/java/co/nimblehq/template/compose/di/modules/StorageModule.kt [hunk 2]
```diff
@@ -17,18 +17,8 @@ abstract class StorageModule {
...
-
         @Provides
         fun provideSecuredLocalStorage(@ApplicationContext context: Context) =
             EncryptedSharedPreferences(context)
-
-        @Provides
-        fun provideNormalLocalStorage(
-            @ApplicationContext context: Context,
-            defaultSharedPreferences: SharedPreferences
-        ) = NormalSharedPreferences(context, defaultSharedPreferences)
```
Comment (luongvo) [inline template-compose/app/src/main/java/co/nimblehq/template/compose/di/modules/StorageModule.kt:32]: same

File: template-compose/data/src/test/java/co/nimblehq/template/compose/data/repository/AppPreferencesRepositoryImplTest.kt [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    private val mockDataStore: DataStore<Preferences> = mockk()
+    private lateinit var repository: AppPreferencesRepository
+
+    @Before
+    fun setup() {
+        repository = AppPreferencesRepositoryImpl(mockDataStore)
+    }
+
+    @Test
+    fun `When calling isFirstTimeLaunch request successfully, it returns correct response`() =
```
Comment (luongvo) [inline template-compose/data/src/test/java/co/nimblehq/template/compose/data/repository/AppPreferencesRepositoryImplTest.kt:47]: we need to update the preference name, for example `When getting a preference` 💭

File: template-compose/data/src/test/java/co/nimblehq/template/compose/data/repository/AppPreferencesRepositoryImplTest.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            repository.getAppPreferences().test {
+                awaitError() shouldBe expected
+            }
+        }
+
+    /**
+     * FIXME Can't use MockK to mock DataStore.edit or DataStore.updateData, switch to use a test DataStore.
+     */
+    @Test
+    fun `When calling updateFirstTimeLaunch, it updates ItSupportInfo into DataStore`() =
```
Comment (luongvo) [inline template-compose/data/src/test/java/co/nimblehq/template/compose/data/repository/AppPreferencesRepositoryImplTest.kt:83]: `When updating a preference` 💭

Comment (luongvo) [inline template-compose/data/src/test/java/co/nimblehq/template/compose/data/repository/AppPreferencesRepositoryImplTest.kt:83]: please replace the `ItSupportInfo` word, it comes from my sample code 🤓

File: template-xml/data/src/test/java/co/nimblehq/template/xml/data/repository/AppPreferencesRepositoryImplTest.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            repository.getAppPreferences().test {
+                awaitError() shouldBe expected
+            }
+        }
+
+    /**
+     * FIXME Can't use MockK to mock DataStore.edit or DataStore.updateData, switch to use a test DataStore.
+     */
+    @Test
+    fun `When calling updateFirstTimeLaunch, it updates ItSupportInfo into DataStore`() =
```
Comment (luongvo) [inline template-xml/data/src/test/java/co/nimblehq/template/xml/data/repository/AppPreferencesRepositoryImplTest.kt:83]: same

File: template-compose/domain/src/main/java/co/nimblehq/template/compose/domain/repository/AppPreferencesRepository.kt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package co.nimblehq.template.compose.domain.repository
+
+import kotlinx.coroutines.flow.Flow
+
+interface AppPreferencesRepository {
+
+    fun getAppPreferences(): Flow<Boolean>
+
+    suspend fun updateAppPreferences(appPreferencesValue: Boolean)
```
Comment (luongvo) [inline template-compose/domain/src/main/java/co/nimblehq/template/compose/domain/repository/AppPreferencesRepository.kt:9]: I think the base name (as the sample) could be just like this

```suggestion
    fun getPreference(): Flow<Boolean>

    suspend fun updatePreference(appPreferenceValue: Boolean)
```

Get or update is an action on a `single preference`, not on all preferences or appPreferences.

File: template-xml/domain/src/main/java/co/nimblehq/template/xml/domain/repository/AppPreferencesRepository.kt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package co.nimblehq.template.xml.domain.repository
+
+import kotlinx.coroutines.flow.Flow
+
+interface AppPreferencesRepository {
+
+    fun getAppPreferences(): Flow<Boolean>
+
+    suspend fun updateAppPreferences(appPreferencesValue: Boolean)
```
Comment (luongvo) [inline template-xml/domain/src/main/java/co/nimblehq/template/xml/domain/repository/AppPreferencesRepository.kt:9]: same

**Review by luongvo — COMMENTED** [review]
The rest lgtm

File: template-xml/app/src/main/java/co/nimblehq/template/xml/di/modules/StorageModule.kt (hunk 1 above)
Comment (lydiasama) [reply template-xml/app/src/main/java/co/nimblehq/template/xml/di/modules/StorageModule.kt:32]: Indeed. Thanks @luongvo

**Comment by lydiasama** [comment]
@luongvo All updated. I also updated the sample repos altogether.

**Review by luongvo — APPROVED** [review]
LGTM

**Review by ryan-conway — APPROVED** [review]

**Review by kaungkhantsoe — APPROVED** [review]

File: sample-compose/app/src/main/java/co/nimblehq/sample/compose/di/modules/PreferencesModule.kt [hunk 8]
```diff
@@ -6,27 +6,26 @@ import androidx.datastore.preferences.core.Preferences
...
-private val Context.appPreferencesDataStore by preferencesDataStore("app_preferences")
 
 @Module
 @InstallIn(SingletonComponent::class)
-class PreferencesModule {
+abstract class PreferencesModule {
+
+    @Binds
+    abstract fun bindAppPreferencesRepository(
+        pppPreferencesRepositoryImpl: AppPreferencesRepositoryImpl
```
Comment (AVI5HEK) [inline sample-compose/app/src/main/java/co/nimblehq/sample/compose/di/modules/PreferencesModule.kt:20]: Here and everywhere below: 😆 
```suggestion
        appPreferencesRepositoryImpl: AppPreferencesRepositoryImpl
```

**Comment by ryan-conway** [comment]
@lydiasama conflicts 🙈

**Review by AVI5HEK — APPROVED** [review]

**Comment by lydiasama** [comment]
@ryan-conway @luongvo All updated.

**Comment by luongvo** [comment]
> @ryan-conway @luongvo All updated.

Looks good 👍

**Pull request merged**